### DISCUSSION TITLE
Window template

### DISCUSSION THREAD

**Discussion by MichaelSvejcar** [AvaloniaUI/Avalonia#10706] [Q&A] [answered]
I'm looking to streamline the creation of windows in my application by implementing a window template. This template would feature a custom top panel with predefined controls, like text boxes and icons, which would be reused across different windows. My goal is to avoid having to specify this part of the window GUI for each window separately.

In the template, I want to define a grid that has two parts. The first part would display the controls that are already defined in the template, such as text boxes. The second part would show any additional controls that are defined in windows that inherit from this template. This way, the controls from the first part of the grid would always be visible.

I'm also looking for a way to implement some properties for the template, such as a string that represents the text in the custom top panel's textbox, bool defining if the panel is visible at all or icon, that will be shown in the top panel.

Could you please advise me on how to implement this? Thanks in advance, and have a great Sunday! ✌️

**Comment by jp2masa** [comment]
There are mainly 2 approaches I would consider:

- (MVVM only) Create the window viewmodel with a property `ViewModel Content { get; }` and properties for the fixed part and then implement the corresponding view with the grid and the corresponding bindings.
- Inherit from window, define Avalonia properties for content and the fixed part, then define a style with a custom template with the grid and the corresponding template bindings.

**Reply by MichaelSvejcar** [reply]
Could you please provide more details on how to implement the first approach? If possible, it would be great if you could include some code snippets or provide a more detailed explanation of each step, as I am still pretty new to Avalonia. Thank you, I appreciate it!

**Reply by lukaszossowski** [reply]
As for how you would go about the first approach, if I understand the suggestion of jp2masa correctly, you would have to create a window with a grid, for example with 2 rows. The first row would have all the constant controls and the second one, where the different content would be would be a ContentControl that is bound to the viewmodels Content property. Then in the ContentControl you can specify datatemplates which will display different view according to the viewmodel that you assign to the Content property.

**Reply by jp2masa** [reply]
Here is an example, the `{ get; set; }` in the viewmodels must be implemented according to the MVVM framework you are using (or INPC):

```cs
class FixedContentViewModel : ViewModel
{
    public string Text1 { get; set; }

    public string Text2 { get; set; }
}
```

```cs
class ExampleViewModel : ViewModel
{
    public string ExampleText { get; set; }
}
```

```cs
class RootViewModel : ViewModel
{
    public RootViewModel(ViewModel content)
    {
        FixedContent = new FixedContentViewModel();
        Content = content;
    }

    public FixedContentViewModel FixedContent { get; }

    public ViewModel Content { get; }
}
```

```xaml
<UserControl
    ...
    x:DataType="vm:RootViewModel">
    <Grid RowDefinitions="Auto, *">
        <ContentControl
            Grid.Row="0"
            Content="{Binding FixedContent}" />
        <ContentControl
            Grid.Row="1"
            Content="{Binding Content}" />
    </Grid>
</UserControl>
```

```xaml
<UserControl
    ...
    x:DataType="vm:FixedContentViewModel">
    <StackPanel>
        <TextBox Text="{Binding Text1}" />
        <TextBox Text="{Binding Text2}" />
    </StackPanel>
</UserControl>
```

```xaml
<UserControl
    ...
    x:DataType="vm:ExampleViewModel">
    <StackPanel>
        <TextBox Text="{Binding ExampleText}" />
    </StackPanel>
</UserControl>
```

**Reply by MichaelSvejcar** [reply]
Thanks for your response! However, I'm little confused here. This way, I can pass multiple view models to one, RootViewModel. But how do I define the part of the window, that I want to be different of each instance of this window template, in the xaml? If I understand correctly, the view itself isn't contained in the view model, that I can then pass to the "parent" view model, or is it?

**Reply by jp2masa** [reply]
The view model is a representation of the view that doesn't depend on the UI framework. The appropriate view would be created from the view model by the view locator (in this case I'm assuming it's in the `DataTemplates`, similar to the MVVM template).

**Comment by lukaszossowski** [comment, marked as answer]
I have done something like that recently, for a dialog window. I needed to have a custom window that I implemented which had a couple of constant elements - a title, X button in the right corner, space for content in center and Cancel/Accept buttons at the bottom. I created a templated control with this codebehind:

```
 public class DialogWindow : Window, IStyleable
    {
        Type IStyleable.StyleKey => typeof(DialogWindow);

        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
        {
            base.OnApplyTemplate(e);
            var button = e.NameScope.Find<Button>("ExitButton");

            if (button != null) button.Click += CancelButton_Click;

            button = e.NameScope.Find<Button>("CancelButton");

            if (button != null) button.Click += CancelButton_Click;

            button = e.NameScope.Find<Button>("AcceptButton");

            if (button != null) button.Click += AcceptButton_Click;
        }

        private void CancelButton_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            Close(false);
        }

        private void AcceptButton_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            Close(true);
        }
    }
```

Which implements a simple behaviour - creates for me a dialog window which returns true or false depending on what the user clicks. Note that it implements the IStyleable interface, without it the style defined in the style part of the templated control wouldn't apply. That is not required in normal templated controls, but seems to be necessary when deriving from Window.

And this is the style part:

```
<Design.PreviewWith>
	<controls:DialogWindow Width="400" Height="400" SizeToContent="Manual" />
</Design.PreviewWith>

<Style Selector="controls|DialogWindow">
		<Setter Property="Background" Value="Transparent" />
		<Setter Property="WindowStartupLocation" Value="CenterOwner"/>
		<Setter Property="TransparencyLevelHint" Value="Transparent"/>
		<Setter Property="SystemDecorations" Value="None"/>
		<Setter Property="ShowInTaskbar" Value="False"/>

		<Setter Property="Template">
			<ControlTemplate>
				<Border CornerRadius="5" BoxShadow="0 0 30 1 #33000000" Margin="30" >
					<Border Background="{StaticResource BackgroundBrush}" CornerRadius="5" ClipToBounds="True">
						<Panel>
							<Grid RowDefinitions="40,*,1,40">
								<Grid >
									<TextBlock Text="{TemplateBinding Title}" HorizontalAlignment="Left" FontSize="12" Margin="20,10" FontWeight="DemiBold" Opacity="0.3"/>

									<Button Name="ExitButton" Width="25" Height="25" Margin="10" Padding="5" HorizontalAlignment="Right">
										<Image Source="/Assets/xBlue.png" />
									</Button>
								</Grid>
								<ContentPresenter Grid.Row="1" Margin="20,0" Content="{TemplateBinding Content}" />

								<Border Grid.Row="2" Background="{StaticResource DataHeaderBrush}" Margin="-20,0" />
								<Grid Grid.Row="3" ColumnDefinitions="*,*" Margin="-20,0">
									<Button Name="CancelButton" Classes="Cancel" FontWeight="Medium" Content="CANCEL" BorderThickness="0" CornerRadius="0" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>

									<Button Name="AcceptButton" Classes="Yellow" FontWeight="Medium" Content="ACCEPT"  BorderThickness="0" CornerRadius="0" Grid.Column="2" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>
								</Grid>
							</Grid>
						</Panel>
					</Border>
				</Border>
			</ControlTemplate>
		</Setter>
	</Style>
```

The important parts here are the buttons, which are named (so that they can be found in the codebehind) and the ContentPresenter which will display the actual content that is inside the window. The preview od this style looks like this:
![window](https://user-images.githubusercontent.com/112421929/226205300-c59b86ea-a41c-444c-881b-e3558b39ae51.JPG)

Now when I want to reuse it I can do this. I create an Avalonia Window and make it inherit from my DialogWindow:

```
 public partial class AddBatchWindow : DialogWindow
    {
        public AddBatchWindow()
        {
            InitializeComponent();
            DataContext = new AddBatchViewModel();
        }
    }
```
Note that you don't have to set the DataContext here, you can do it anywhere you want, or pass it as a consructor etc. Now the XAML looks like this:

```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
       (my project namespaces here)
	Width="400"
	SizeToContent="Height"
        Title="ADD BATCH">

	<StackPanel Spacing="20" Margin="20">
(...)
	</StackPanel>
</Window>
```
Note that there are no accept/cancel or close buttons, I am just creating a stackpanel of a few controls. The result looks like this:

![window2](https://user-images.githubusercontent.com/112421929/226205545-b2a2f38c-f160-4533-8ffc-821c566dfc6c.JPG)

**Reply by MichaelSvejcar** [reply]
Hello, thank you for your response. I have tried to create simple test project with the code you provided. However, I got one issue. I am getting error "Unable to resolve type DialogWindow from namespace WindowTemplateTest.Views".. However, the namespaces seems to be correct. I know it must be some rookie mistake, as I am still pretty new to this. Can you perhaps try to solve it with me, or provide me with some simple, working project containing the functionality you described above? I would be very grateful. This is the exact code on my side:

**DialogWindow.axaml.cs:**
```
namespace WindowTemplateTest.Views
{
    public partial class DialogWindow : Window, IStyleable
    {

        public DialogWindow()
        {

        }

        Type IStyleable.StyleKey => typeof(DialogWindow);

        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
        {
            base.OnApplyTemplate(e);
            var button = e.NameScope.Find<Button>("ExitButton");

            if (button != null) button.Click += CancelButton_Click;

            button = e.NameScope.Find<Button>("CancelButton");

            if (button != null) button.Click += CancelButton_Click;

            button = e.NameScope.Find<Button>("AcceptButton");

            if (button != null) button.Click += AcceptButton_Click;
        }

        private void CancelButton_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            Close(false);
        }

        private void AcceptButton_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            Close(true);
        }
    }
```

And this is the **DialogWindow.axaml:**
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:views="WindowTemplateTest.Views"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="WindowTemplateTest.Views.DialogWindow"
        Title="DialogWindow">

	<Design.PreviewWith>
		<views:DialogWindow Width="400" Height="400" SizeToContent="Manual" />
	</Design.PreviewWith>

	<Style Selector="views|DialogWindow">
		<Setter Property="Background" Value="Transparent" />
		<Setter Property="WindowStartupLocation" Value="CenterOwner"/>
		<Setter Property="TransparencyLevelHint" Value="Transparent"/>
		<Setter Property="SystemDecorations" Value="None"/>
		<Setter Property="ShowInTaskbar" Value="False"/>

		<Setter Property="Template">
			<ControlTemplate>
				<Border CornerRadius="5" BoxShadow="0 0 30 1 #33000000" Margin="30" >
					<Border Background="{StaticResource BackgroundBrush}" CornerRadius="5" ClipToBounds="True">
						<Panel>
							<Grid RowDefinitions="40,*,1,40">
								<Grid >
									<TextBlock Text="{TemplateBinding Title}" HorizontalAlignment="Left" FontSize="12" Margin="20,10" FontWeight="DemiBold" Opacity="0.3"/>

									<Button Name="ExitButton" Width="25" Height="25" Margin="10" Padding="5" HorizontalAlignment="Right">
										<Image Source="/Assets/xBlue.png" />
									</Button>
								</Grid>
								<ContentPresenter Grid.Row="1" Margin="20,0" Content="{TemplateBinding Content}" />

								<Border Grid.Row="2" Background="{StaticResource DataHeaderBrush}" Margin="-20,0" />
								<Grid Grid.Row="3" ColumnDefinitions="*,*" Margin="-20,0">
									<Button Name="CancelButton" Classes="Cancel" FontWeight="Medium" Content="CANCEL" BorderThickness="0" CornerRadius="0" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>

									<Button Name="AcceptButton" Classes="Yellow" FontWeight="Medium" Content="ACCEPT"  BorderThickness="0" CornerRadius="0" Grid.Column="2" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" VerticalContentAlignment="Center" HorizontalContentAlignment="Center"/>
								</Grid>
							</Grid>
						</Panel>
					</Border>
				</Border>
			</ControlTemplate>
		</Setter>
	</Style>
	
	
</Window>
```

**Reply by lukaszossowski** [reply]
I see one difference here - you have a public partial class DialogWindow. I am assuming you created a new Avalonia Window, as it creates a partial class. Please create a Templated Control. It should create both axaml.cs and axaml file, but the axaml.cs should have a regular class, not partial. Creating a window file also creates some autogenerated code which probably is the issue.

**Reply by MichaelSvejcar** [reply]
Thanks, I have resolved this issue. However, I still have problem - I created new Templated Control, representing the window, which then inherits from the DialogWindow. But no matter what controls I define in it's xaml, the preview is just blank, same as when I run the app. I am not sure if that is because of the DataContext setup, but how do I make sure, that the controls defined in the xaml of the new TemplatedControl, will get "passed" to the DialogWindow's ContentPresenter, so the whole window gets shown up?  Thanks!

**Reply by lukaszossowski** [reply]
For any templated control to actually show the defined style you have to include it in the App.axaml file:
```
	<Application.Styles>
         (...)
	          <StyleInclude Source="/Controls/DialogWindow.axaml" />
	</Application.Styles>
```

As far as I can tell you don't need to include styles for the derived windows (like my AddBatchWindow in the example), just the main DialogWindow.

The ContentPresenter's Content property is bound to the Content property of the DialogWindow. In effect, content is whatever you put between the Window tags in the xaml file.

**Reply by timunie** [reply]
check out the `:is` selector. It may help in this specific case, see: https://docs.avaloniaui.net/docs/styling/selectors#is

**Reply by MichaelSvejcar** [reply]
Including the style helped, now I am able to see the preview of the inherited window in a way one would expect. Appreciate your help @lukaszossowski, thank you!

**Reply by MichaelSvejcar** [reply]
BTW, how would I implement some property of this window? Let's say I want bool property IsMinimizeButtonVisible, which will toggle minimize button visibility. I also want to be able to set this property in the xaml of the window that inherits this control. How do I do that?

**Reply by lukaszossowski** [reply]
Have you tried creating a styled property like you would on any other custom control?

**Reply by MichaelSvejcar** [reply]
Yes, but I couldn't see it or set it in any window that inherited from this one.

**Reply by timunie** [reply]
@MichaelSvejcar https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample#step-3-add-the-numberofstars-property for example

**Reply by MichaelSvejcar** [reply]
Hey guys, do you know how to make this work on  Avalonia 11? I tried updating and checking the breaking changes list, but it seems that this area is not documented enough. Thank you.

**Reply by timunie** [reply]
@MichaelSvejcar can you specify what doesn't work for you exactly?

**Reply by MichaelSvejcar** [reply]
@timunie I was playing with the update a little bit more and so far, I have successfully ported almost everything. 
One thing related to this topic that I was not able to port so far is my CustomWindow, which inherits from Window. I am using this window as a template for all the windows in my app, which are inheriting it. In it's xaml, it is structurized like this:

```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
		xmlns:views="using:CustomProject_Avalonia.Views">
	
	<Design.PreviewWith>
		<views:CustomWindow Width="400" Height="400" SizeToContent="Manual" />
	</Design.PreviewWith>

	<StyleInclude Source="/Styles/Styles.axaml"></StyleInclude>

	<Style Selector="views|CustomWindow">
		<Setter Property="Background" Value="Transparent" />
		<Setter Property="WindowStartupLocation" Value="CenterOwner"/>
		<Setter Property="TransparencyLevelHint" Value="Transparent"/>
		<Setter Property="SystemDecorations" Value="Full"/>
		<Setter Property="ShowInTaskbar" Value="False"/>

		<Setter Property="Template">
			<ControlTemplate>
				<Grid>
					<Grid.RowDefinitions>
						<RowDefinition Height="Auto"/>
						<RowDefinition Height="*"/>
					</Grid.RowDefinitions>
					
					<!--Window body-->
					<Border Grid.Row="0" Grid.RowSpan="2" BorderThickness="0.5" Padding="0" Background="#020202">
						<Border BorderThickness="3" Padding="2"	Background="#202020">
							<Grid Background="#202020">
								<Grid.RowDefinitions>
									<RowDefinition Height="30"/>
									<RowDefinition Height="3"/>
									<RowDefinition Height="*"/>
								</Grid.RowDefinitions>

								<!--Top window panel grid-->
								<Grid Name="TopPanel" Grid.Row="0" ShowGridLines="False" Background="Transparent">
									<Grid.ColumnDefinitions>
										<ColumnDefinition Width="30"></ColumnDefinition>
										<ColumnDefinition Width="*"></ColumnDefinition>
										<ColumnDefinition Width="Auto"></ColumnDefinition>
									</Grid.ColumnDefinitions>

									<Border Grid.Column="0">
										<Svg Path="{TemplateBinding	IconSvg}" Height="25"/>
									</Border>

									<Border Grid.Column="1" BorderThickness="4">
										<TextBlock Text="{TemplateBinding Title}" Foreground="White" VerticalAlignment="Center" FontWeight="Bold" FontSize="15"/>
									</Border>

									<!--Close, minimize, maximize buttons-->
									<StackPanel Grid.Column="2" Orientation="Horizontal" VerticalAlignment="Center">
										<Button  Name="MinimizeButton" Background="Transparent" Width="30" Height="30" Padding="0" Margin="0">
											<Button.Content>
												<Svg Width="15" Path="/Assets/ButtonImages/window-minimize.svg"></Svg>
											</Button.Content>
										</Button>
										<Button Name="MaximizeButton" Background="Transparent" Width="30" Height="30" Padding="0" Margin="0">
											<Button.Content>
												<Panel>
													<Panel Name="SvgMaximizePanel">
														<Svg Name="SvgMaximize" Width="15" Path="/Assets/ButtonImages/window-maximize.svg"/>
													</Panel>
													<Panel Name="SvgRestorePanel">
														<Svg Name="SvgRestore" Width="15" Path="/Assets/ButtonImages/window-restore.svg"/>
													</Panel>
												</Panel>
											</Button.Content>
										</Button>
										<Button Name="CloseButton" Background="Transparent" Width="30" Height="30" Padding="0" Margin="0">
											<Button.Content>
												<Svg Width="15" Path="/Assets/ButtonImages/window-close.svg"/>
											</Button.Content>
										</Button>
										
									</StackPanel>
								</Grid>

								<!--Content-->
								<Border Grid.Row="2" ClipToBounds="True">
									<Panel>
										<Grid>
											<Grid.RowDefinitions>
												<RowDefinition Height="*"/>
												<RowDefinition Height="3"/>
												<RowDefinition Height="30"/>
											</Grid.RowDefinitions>
											<Border Grid.Row="0" CornerRadius="3" Background="#404040">
											</Border>
										</Grid>
										<ContentPresenter Content="{TemplateBinding Content}"/>
									</Panel>
								</Border>
							</Grid>
						</Border>
					</Border>
				</Grid>
			</ControlTemplate>
		</Setter>
	</Style>
</Styles>
```
The problem is that in it's code behind in the OnApplyTemplate, I was initializing events for it's controls through their names as defined in the xaml using this.FindControl<>(), like this:

```
protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
{
    SystemDecorations = SystemDecorations.BorderOnly;
    WindowStartupLocation = WindowStartupLocation.CenterOwner;

    InitializeIfNeeded();

    // Initialize events
    base.OnApplyTemplate(e);
    this.KeyDown += CustomWindow_KeyDown;
    // popup related events
    this.AddHandler(PointerPressedEvent, CustomWindow_PointerPressed, handledEventsToo: true);
    InitializePopups();

    //initializes top panel dragging
    var topPanel = this.FindControl<Grid>("TopPanel");
    topPanel.PointerPressed += MoveWindowElementPressed;

    // first assign the events
    var minimizeButton = e.NameScope.Find<Button>("MinimizeButton");
    if (minimizeButton != null) minimizeButton.Click += MinimizeButton_Click;

    var maximizeButton = e.NameScope.Find<Button>("MaximizeButton");
    if (maximizeButton != null) maximizeButton.Click += MaximizeButton_Click;

    var closeButton = e.NameScope.Find<Button>("CloseButton");
    if (closeButton != null) closeButton.Click += CloseButton_Click;

    // set maximize, minimize visibilities
    minimizeButton.IsVisible = IsMinimizeAndMaximizeVisible;
    maximizeButton.IsVisible = IsMinimizeAndMaximizeVisible;

    // find panels containing maximize/restore svgs in maximize buttons
    var svgMaximizePanel = e.NameScope.Find<Panel>("SvgMaximizePanel");
    var svgRestorePanel = e.NameScope.Find<Panel>("SvgRestorePanel");

    // Subscribe to the WindowStateObservable
    WindowStateObservable.Subscribe(state =>
    {
        if (state == WindowState.Maximized)
        {
            // The window is maximized, so show the restore button and hide the maximize button
            svgMaximizePanel.IsVisible = false;
            svgRestorePanel.IsVisible = true;
        }
        else if (state == WindowState.Normal)
        {
            // The window is not maximized, so show the maximize button and hide the restore button
            svgMaximizePanel.IsVisible = true;
            svgRestorePanel.IsVisible = false;
        }
    });
}
```

However, the problem is that in Avalonia 11, using this.Find<>() is not used anymore. Instead of that, same-named properties as the control names in view are automatically generated in the code behind, to reference these controls. The problem is that this is not really a Window but a Styles instead, so I cannot reference the controls in the code behind neither of these ways anymore.
What is the solution to this problem?

Thanks in advance,
Michael

**Reply by timunie** [reply]
in OnApplyTemplate, you can use `e.Namespace.Find` and `e.Namespace.Get<TargetType>`.

**Reply by H0tCh0colat3** [reply]
To add to this for anyone who stumbles upon this answer in the future, if you want to use flyouts or adorners in a re-templated window, make sure you wrap the content in the template with a VisualLayerManager. [See the window template that fluent theme uses](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Window.xaml). VisualLayerManager handles the LightDismissOverlayLayer, which flyout seems to require in order to know when to dismiss.
